File: src/test/java/edu/kit/iti/formal/stvs/TestUtils.java
package edu.kit.iti.formal.stvs;

import edu.kit.iti.formal.stvs.logic.io.ImportException;
import edu.kit.iti.formal.stvs.logic.io.ImporterFacade;
import edu.kit.iti.formal.stvs.model.common.FreeVariableList;
import edu.kit.iti.formal.stvs.model.common.FreeVariableListValidator;
import edu.kit.iti.formal.stvs.model.common.FreeVariableProblem;
import edu.kit.iti.formal.stvs.model.common.ValidFreeVariable;
import edu.kit.iti.formal.stvs.model.config.GlobalConfig;
import edu.kit.iti.formal.stvs.model.expressions.Type;
import edu.kit.iti.formal.stvs.model.expressions.TypeBool;
import edu.kit.iti.formal.stvs.model.expressions.TypeEnum;
import edu.kit.iti.formal.stvs.model.expressions.TypeInt;
import edu.kit.iti.formal.stvs.model.table.ConstraintSpecification;
import edu.kit.iti.formal.stvs.model.table.ValidSpecification;
import edu.kit.iti.formal.stvs.model.table.problems.ConstraintSpecificationValidator;
import edu.kit.iti.formal.stvs.model.table.problems.SpecProblem;
import javafx.beans.property.SimpleObjectProperty;
import javafx.collections.FXCollections;
import javafx.stage.FileChooser;
import org.junit.Assume;
import org.mockito.Mockito;
import org.mockito.stubbing.OngoingStubbing;
import org.powermock.api.mockito.PowerMockito;

import java.io.File;
import java.io.InputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static org.mockito.ArgumentMatchers.any;
import static org.testfx.util.WaitForAsyncUtils.sleep;

/**
 * Created by bal on 10.02.17.
 */
public class TestUtils {

  /**
   * Tolerance for floating-point rounding errors when doing assertEquals() with doubles
   */
  public static final double EPSILON = 0.001;

  public static ValidSpecification importValidSpec(InputStream source, TypeEnum... enumTypes) {
    List<Type> typeContext = new ArrayList<>();
    typeContext.add(TypeInt.INT);
    typeContext.add(TypeBool.BOOL);
    for (TypeEnum enumType : enumTypes) {
      typeContext.add(enumType);
    }
    return importValidSpec(source, typeContext);
  }

  public static ValidSpecification importValidSpec(
      InputStream source,
      List<Type> typeContext) {
    try {
      ConstraintSpecification constraintSpec = ImporterFacade.importConstraintSpec(source, ImporterFacade.ImportFormat.XML);
      List<ValidFreeVariable> validFreeVariables = importValidFreeVariables(
          constraintSpec.getFreeVariableList(), typeContext);
      ConstraintSpecificationValidator validator = new ConstraintSpecificationValidator(
          new SimpleObjectProperty<>(typeContext),
          new SimpleObjectProperty<>(new ArrayList<>()),
          new SimpleObjectProperty<>(validFreeVariables),
          constraintSpec
      );
      ValidSpecification validSpec = validator.getValidSpecification();
      if (validSpec == null) {
        System.err.println("ConstraintSpecification could not be validated:");
        validator.problemsProperty().get().stream().map(SpecProblem::getErrorMessage).forEach(System.err::println);
        throw new RuntimeException("Couldn't validate Specification");
      }
      return validSpec;
    } catch (ImportException cause) {
      throw new RuntimeException(cause);
    }
  }

  public static List<ValidFreeVariable> importValidFreeVariables(InputStream source, TypeEnum... enumTypes) {
    List<Type> typeContext = new ArrayList<>();
    typeContext.add(TypeInt.INT);
    typeContext.add(TypeBool.BOOL);
    for (TypeEnum enumType : enumTypes) {
      typeContext.add(enumType);
    }
    try {
      return importValidFreeVariables(
          ImporterFacade.importConstraintSpec(source, ImporterFacade.ImportFormat.XML).getFreeVariableList(),
          typeContext);
    } catch (ImportException ex) {
      throw new RuntimeException(ex);
    }
  }

  public static List<ValidFreeVariable> importValidFreeVariables(FreeVariableList freeVariableList, List<Type> typeContext) {
    FreeVariableListValidator validator = new FreeVariableListValidator(
        new SimpleObjectProperty<>(typeContext), freeVariableList);
    if (!validator.problemsProperty().get().isEmpty()) {
      System.err.println("Could not validate free variables:");
      validator.problemsProperty().get().entrySet().stream()
          .map(entry -> entry.getKey() + " -> "
              + entry.getValue().stream()
              .map(FreeVariableProblem::getGuiMessage)
              .collect(Collectors.toList()))
          .forEach(System.err::println);
      throw new RuntimeException("Couldn't validate");
    }
    return validator.validFreeVariablesProperty().get();
  }

  public static void gimmeTime() {
    sleep(5, TimeUnit.SECONDS);
  }

  public static void mockFiles(URL ... urls) throws Exception {
    List<File> files = Arrays.stream(urls)
        .map(URL::getPath)
        .map(File::new)
        .collect(Collectors.toList());
    FileChooser chooser = Mockito.mock(FileChooser.class);
    OngoingStubbing<File> stub = Mockito.when(chooser.showOpenDialog(any()));
    for (File file : files) {
      stub = stub.thenReturn(file);
    }
    Mockito.when(chooser.getExtensionFilters()).thenReturn(FXCollections.observableList(new ArrayList<>()));
    PowerMockito.whenNew(FileChooser.class).withAnyArguments().thenReturn(chooser);
  }

    private static void assumeFileExists(String message, String executable) {
        Assume.assumeTrue(message, new File(executable).exists());
    }

    public static void assumeZ3Exists() {
        assumeFileExists(
                "The z3 command is not set or a non-existing file. Tests are skipped!",
                GlobalConfig.autoloadConfig().getZ3Path());
    }

    public static void assumeNuXmvExists() {
        assumeFileExists(
                "The nuxmv command is not set or a non-existing file. Tests are skipped!",
                GlobalConfig.autoloadConfig().getNuxmvFilename());
    }

    public static void assumeGetetaExists() {
        String[] toks = GlobalConfig.autoloadConfig().getGetetaCommand()
                .split(" ");
        assumeFileExists(
                "The geteta command is not set or a non-existing file. Tests are skipped!",
                toks[0]);
    }
}

File: stvs/src/test/kotlin/edu/kit/iti/formal/stvs/TestUtils.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs

import edu.kit.iti.formal.stvs.logic.io.ImportException
import edu.kit.iti.formal.stvs.logic.io.ImporterFacade
import edu.kit.iti.formal.stvs.model.code.Code
import edu.kit.iti.formal.stvs.model.code.CodeTest
import edu.kit.iti.formal.stvs.model.common.FreeVariableList
import edu.kit.iti.formal.stvs.model.common.FreeVariableListValidator
import edu.kit.iti.formal.stvs.model.common.FreeVariableProblem
import edu.kit.iti.formal.stvs.model.common.ValidFreeVariable
import edu.kit.iti.formal.stvs.model.config.GlobalConfig.Companion.autoloadConfig
import edu.kit.iti.formal.stvs.model.expressions.Type
import edu.kit.iti.formal.stvs.model.expressions.TypeBool
import edu.kit.iti.formal.stvs.model.expressions.TypeEnum
import edu.kit.iti.formal.stvs.model.expressions.TypeInt
import edu.kit.iti.formal.stvs.model.table.ConstraintSpecification
import edu.kit.iti.formal.stvs.model.table.ValidSpecification
import edu.kit.iti.formal.stvs.model.table.problems.ConstraintSpecificationValidator
import javafx.beans.property.SimpleListProperty
import javafx.collections.FXCollections
import org.junit.Assume
import tornadofx.asObservable
import java.io.File
import java.io.InputStream
import java.io.InputStreamReader

/**
 * Created by bal on 10.02.17.
 */
object TestUtils {
    /**
     * Tolerance for floating-point rounding errors when doing assertEquals() with doubles
     */
    const val EPSILON: Double = 0.001

    fun loadCodeFromFile(filename: String): Code =
        InputStreamReader(CodeTest::class.java.getResourceAsStream(filename)!!).use {
            Code(filename, it.readText())
        }

    fun loadFromTestSets(resource: String): InputStream =
        TestUtils::class.java.getResourceAsStream("testSets/$resource")!!

    fun importValidSpec(source: InputStream, vararg enumTypes: TypeEnum): ValidSpecification {
        val typeContext: MutableList<Type> = ArrayList()
        typeContext.add(TypeInt)
        typeContext.add(TypeBool)
        for (enumType in enumTypes) {
            typeContext.add(enumType)
        }
        return importValidSpec(source, typeContext)
    }

    fun importValidSpec(source: InputStream, typeContext: List<Type>): ValidSpecification {
        try {
            val constraintSpec: ConstraintSpecification =
                ImporterFacade.importConstraintSpec(source)
            val validFreeVariables = importValidFreeVariables(
                constraintSpec.freeVariableList,
                typeContext,
            )
            val validator = ConstraintSpecificationValidator(
                SimpleListProperty(typeContext.asObservable()),
                SimpleListProperty(FXCollections.observableArrayList()),
                validFreeVariables.asObservable(),
                constraintSpec,
            )
            val validSpec = validator.validSpecification
            if (validSpec == null) {
                System.err.println("ConstraintSpecification could not be validated:")
                validator.problemsProperty.get().map { it?.errorMessage }.forEach { System.err.println(it) }
                throw RuntimeException("Couldn't validate specification")
            }
            return validSpec
        } catch (cause: ImportException) {
            throw RuntimeException(cause)
        }
    }

    fun importValidFreeVariables(source: InputStream, vararg enumTypes: TypeEnum): List<ValidFreeVariable> {
        val typeContext: MutableList<Type> = ArrayList()
        typeContext.add(TypeInt)
        typeContext.add(TypeBool)
        for (enumType in enumTypes) {
            typeContext.add(enumType)
        }
        try {
            return importValidFreeVariables(
                ImporterFacade.importConstraintSpec(source).freeVariableList,
                typeContext,
            )
        } catch (ex: ImportException) {
            throw RuntimeException(ex)
        }
    }

    fun importValidFreeVariables(freeVariableList: FreeVariableList, typeContext: List<Type>): List<ValidFreeVariable> {
        val validator = FreeVariableListValidator(SimpleListProperty(typeContext.asObservable()), freeVariableList)
        if (validator.problemsProperty.get().isNotEmpty()) {
            System.err.println("Could not validate free variables:")
            validator.problemsProperty.get().map { (k, v) ->
                "$k -> " + v.map(FreeVariableProblem::guiMessage)
            }.forEach(System.err::println)
            throw RuntimeException("Couldn't validate")
        }
        return validator.validFreeVariablesProperty.get()
    }

    private fun assumeFileExists(message: String, executable: String) {
        Assume.assumeTrue(message, File(executable).exists())
    }

    fun assumeZ3Exists() {
        assumeFileExists(
            "The z3 command is not set or a non-existing file. Tests are skipped!",
            autoloadConfig().z3Path,
        )
    }

    fun assumeNuXmvExists() {
        assumeFileExists(
            "The nuxmv command is not set or a non-existing file. Tests are skipped!",
            autoloadConfig().nuxmvFilename,
        )
    }
}